linker/Linker/BodyRequirementsCollector.cs

```diff
@@ -0,0 +1,119 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Mono.Linker {
+	public static class BodyRequirementsCollector {
+		public static IEnumerable<InterfaceImplementation> InterfacesNeededByBody (AnnotationStore annotations, MethodBody body)
+		{
+			var interfaceImplementations = new HashSet<InterfaceImplementation> ();
+			var possibleStackTypes = AllPossibleStackTypes (body.Method);
+			if (possibleStackTypes.Count == 0)
+				return interfaceImplementations;
+
+			var interfaceTypes = possibleStackTypes.Where (t => t.IsInterface).Distinct ().ToArray ();
+			if (interfaceTypes.Length == 0)
+				return interfaceImplementations;
+
+			// We only sweep interfaces on classes so that's why we only care about classes
+			var classes = possibleStackTypes.Where (t => t.IsClass).Distinct ().ToArray ();
+			if (classes.Length == 0)
+				return interfaceImplementations;
+
+			// If a type could be on the stack in the body and an interface it implements could be on the stack on the body
```

Comment:
I am not sure this will cover all cases (will have to experiment with it)

Reply:
One thing we don't have tests for, and I thought this might not be enough to address is covariance and contravariance.

If there is a bug with covariance and/or contravariance I'd rather address it in a follow on pr.  That is an increasingly less common scenario where this interface sweep bug could appear.